src/rust/lib/graph-editor/src/component/visualization/fullscreen.rs

```diff
@@ -0,0 +1,338 @@
+                animation_data.apply_interpolation(value);
+            },
+            _  => ()
+        }
+        // Check for end of animation and update the state.
+        if value >= 1.0 {
+           state.borrow_mut().animation_end_transition();
+        }
+    }
+}
+
+// ==================
+// === StateModel ===
+// ==================
+
+/// Represents the internal state of the FullscreenState. This is used to ensure animations always
+/// finish without interruption.
+#[derive(Debug,Clone)]
+enum StateModel<T> {
+    /// There is a UI component and it is in fullscreen mode.
+    Fullscreen {
+        data           : FullscreenStateData<T>,
+        resize_handle  : callback::Handle,
+    },
+    /// There is an animation running from fullscreen mode to non-fullscreen mode.
+    TransitioningFromFullscreen {
+        animation_data : AnimationTargetState<T>
+    },
+    /// There is an animation running from non-fullscreen mode to fullscreen mode.
+    TransitioningToFullscreen {
+        animation_data : AnimationTargetState<T>,
+        target_state   : FullscreenStateData<T>
+    },
+    /// There is no UI component in fullscreen mode.
+    NotFullscreen
+}
+
+impl<T:Fullscreenable> StateModel<T> {
+    /// Called to indicate the the running animation has ended. Changes the state to the correct
+    /// follow up state. Does nothing if no animation was running.
+    fn animation_end_transition(&mut self) {
+        let new_state = match self.clone() {
+            StateModel::TransitioningFromFullscreen { .. } => {
+                StateModel::NotFullscreen
+            }
+            StateModel::TransitioningToFullscreen { target_state, ..} => {
+                let resize_handle = target_state.make_resize_handle();
+                StateModel::Fullscreen { data: target_state, resize_handle }
+            }
+            other    => other,
+        };
+        *self = new_state;
+    }
+
+    /// Returns whether there is a component that is in fullscreen mode. Animation phases count as
+    /// still in fullscreen mode.
+    pub fn is_fullscreen(&self) -> bool {
+        match self {
+            StateModel::NotFullscreen{..} => false,
+            _                             => true,
+        }
+    }
+}
+
+impl<T> Default for StateModel<T> {
+    fn default() -> Self {
+        StateModel::NotFullscreen
+    }
+}
+
+
+
+// =======================
+// === FullscreenState ===
+// =======================
+
+/// The `FullscreenState` manages the state changes between fullscreen mode and non-fullscreen mode
+/// for a UI component. It creates animations for the state changes and ensure that the component
+/// cannot come into an illegal state during the transition.
+#[derive(Debug,CloneRef,Derivative)]
+#[derivative(Clone(bound=""))]
+pub struct FullscreenState<T> {
+    network   : frp::Network,
+    state     : Rc<RefCell<StateModel<T>>>,
+    animation : DynSimulator<f32>,
```

Comment:
Please code all UI animations using FRP animations system only. Doing that using raw DynSimulator leads to a very complex code (like here). See for example animations in `cursor.rs`.